src/bmson/parse.rs

```diff
@@ -0,0 +1,903 @@
+    if let Some(array) = value.as_array() {
+        let mut events = Vec::new();
+        for item in array {
+            if let Some(obj) = item.as_object() {
+                let x = if let Some(x_value) = obj.get("x") {
+                    if x_value.is_null() {
+                        None
+                    } else if let Some(n) = x_value.as_u64() {
+                        if n == 0 {
+                            None
+                        } else if n <= u8::MAX as u64 {
+                            std::num::NonZeroU8::new(n as u8)
+                        } else {
+                            warnings.push(BmsonWarning::InvalidFieldValue(format!(
+                                "Key x value {} is too large",
+                                n
+                            )));
+                            None
+                        }
+                    } else {
+                        warnings.push(BmsonWarning::InvalidFieldType(
+                            "Key x should be a number or null".to_string(),
+                        ));
+                        None
+                    }
+                } else {
+                    None
+                };
+                let y = get_u64_field(obj, "y", 0, warnings);
+                events.push(super::KeyEvent {
+                    x,
+                    y: PulseNumber(y),
+                });
+            } else {
+                warnings.push(BmsonWarning::InvalidFieldType(
+                    "Key event should be an object".to_string(),
+                ));
+            }
+        }
+        events
+    } else {
+        warnings.push(BmsonWarning::InvalidFieldType(
+            "Key events should be an array".to_string(),
+        ));
+        Vec::new()
+    }
+}
+
+/// Create a default BMSON object.
+fn create_default_bmson(warnings: &mut Vec<BmsonWarning>) -> Bmson {
+    Bmson {
+        version: "1.0.0".to_string(),
+        info: create_default_bmson_info(warnings),
+        lines: None,
+        bpm_events: Vec::new(),
+        stop_events: Vec::new(),
+        sound_channels: Vec::new(),
+        bga: super::Bga::default(),
+        scroll_events: Vec::new(),
+        mine_channels: Vec::new(),
+        key_channels: Vec::new(),
+    }
+}
+
+/// Create a default BMSON info object.
+fn create_default_bmson_info(warnings: &mut Vec<BmsonWarning>) -> BmsonInfo {
+    BmsonInfo {
+        title: "Unknown".to_string(),
+        subtitle: String::new(),
+        artist: "Unknown".to_string(),
+        subartists: Vec::new(),
+        genre: "Unknown".to_string(),
+        mode_hint: "beat-7k".to_string(),
+        chart_name: String::new(),
+        level: 1,
+        init_bpm: FinF64::new(120.0).unwrap_or_else(|| {
+            warnings.push(BmsonWarning::InvalidFieldValue(
+                "Invalid init_bpm, using default 120.0".to_string(),
+            ));
+            FinF64::new(120.0).unwrap()
+        }),
```

Comment:
```suggestion
        init_bpm: FinF64::new(120.0).unwrap(),
```
It seems that lacks self-consistency. You are using `unwrap` for `FinF64::new(120.0)`, but it is a handling error of `FinF64::new(120.0)`. So if this code works, the warning will not occur.